### ISSUE TITLE
rendering of large amount of markers

### ISSUE THREAD

**Issue by yeedle** [xkjyeah/vue-google-maps#178]
This can be an issue with google maps itself too, but I have found that it is more pronounced with vue-google-maps. Loading more than 5k markers will cause the webpage to slow down significantly or even freeze it completely. I added marker clustering, thinking that would somewhat alleviate the issue, but it seems like every marker gets rendered first, and only after it is finished rendering, the markers are removed from the map and added to the cluster markers. 

Is there a way around this? If my `<gmap-markers>`s are wrapped in a `gmap-cluster>` tag is it possible to not render the markers at all instead of rendering them and then removing them? And more generally, is there any gotchas to avoid when using vue-google-maps that can cause performance improvements?

**Comment by martijnhiemstra** [comment]
I can confirm this. In my case I have about 2000 markers and the first time they will load. I initialize the markers with an array in data like this:

    export default {
        data () {
            return {
                markers: []

In my case I click on a button and that loads about 2000 markers into the marker object. Then when I click on the button the map goes completly haywire. I see really strange behaviour. Either th ebrowser hanging or even markers jumping from place to place randomly when I move the mouse.

**Edit:**
I have justed removed gmap-cluster and now it works fine. I have an idee why it's going wrong.

When the markers array is emptied or changed then the markers get updated on the screen. That triggers the update/redraw of the map that causes a kind of loop or something like that. 

I am guessing that there is a watcher on the markers. If so then the updating/redrawing of the map should be disabled while updating the markers.

**Comment by xkjyeah** [comment]
I did some profiling recently.

Half the time was spent in GC overheads while creating the DOM comment
nodes that stub the Vue component, and half the time spent was initializing
the components.

Unfortunately both are out of my control. They are overheads of the Vue
framework.

One solution is to access the Google Maps API directly. I'll leave that as
an exercise for the reader.

The other is to have an auxiliary component (c.f. ui-gmap-markers from the
angular library) that iterates over the data and creates the markers for
you. This method, while it works, is equally awkward because you end up
passing in strings/callbacks instead of referencing properties directly,
and the API ends up diverging from Google's.

In short, please use the first option. If you think the second option makes
sense, raise a PR. I'll merge it into a library of "contributed" components
with the warning that it doesn't correspond to anything in the Google Maps
JS API

**Comment by yeedle** [comment]
Thanks, that makes a lot of sense. 

Re the first option, do you mean through `$refs.map.$mapObject`? I have run into issues with that from time to time, depending on where I tried to call it (for example, creating an infowindow with the Google Maps api directly requires a reference to a map and a marker, and if I created the infowindow inside a click handler, I would sometimes get a `$markerObject is undefined` error). Where do you suggest is the best place to do stuff via the Google Maps api directly? In the `$refs.map.loaded.then()` promise? Or should I do everything inside a `@tiles_loaded` listener? Or perhaps somewhere else? Would appreciate to hear your opinion.

**Comment by nick-falco** [comment]
I was also running into issues. The solution for me was to use PNG images for markers instead of drawing them with SVG. I am now able to plot over 5000 points without issue.

**Comment by wasinr** [comment]
hi, can I ask you one question.

my code is 

```
<gmap-map :center="center" ref="mmm" :zoom="12" @dragend="draggedMapEnd">
    <gmap-marker v-for="m in markers" :position="m.position"></gmap-marker>
  </gmap-map>
```

after I load markers from http, it doesn't show marker on the map, how do I fix it.

Thank you

**Comment by dorianamouroux** [comment]
I implemented by accessing the Google Map object directly: 

```
<template>
  <div>
    <gmap-map
      :center="{lat: 0, lng: 0}"
      :zoom="2"
      ref="map"
      style="width: 100%; height: 800px">
    </gmap-map>
  </div>
</template>
<script>
import MarkerClusterer from 'marker-clusterer-plus';

export default {
  props: [
    'elems'
  ],
  mounted() {
    this.$nextTick(() => {
      this.$refs.map.$mapCreated.then(() => {
        // wait until the map is mounted
        this.cluster = new MarkerClusterer(this.$refs.map.$mapObject, [], {});
        this.cluster.addMarkers(this.elems);
      })
    })
  },

  watch: {
    elems(old, newValue) {
      this.cluster.clearMarkers();
      this.cluster.addMarkers(newValue);
    }
  }
</script>
```

It is way faster, but it's still not 100% fully responsive. I think using Vue.Js component to represent a Marker is too expensive, it would be better if the markers were passed as props.

**Comment by JiaweiTan** [comment]
@xkjyeah Regarding the solution to access the Google Maps API directly for clustering, is it possible to use vue2 google maps library with it?  The reason is because I have done a lot using this library and was wondering if changing the clustering to google maps api will have major changes to my code.

**Comment by arenarium-dev** [comment]
Hey I dont know if its still relevant but I've been building a plugin for google maps and other map providers. It organizes your map markers and displays them in a way to maximize readability and clarity. It also improves the performance of displaying a large number of markers.

When using it with vue you would mount the bodies of markers with something like createVNode and render.

![Image](https://github.com/user-attachments/assets/dcf83a50-e9db-4132-b798-95009c97bd59)

If you want check it out: [https://arenarium.dev/?demo=bookings](https://arenarium.dev/?demo=bookings)
